flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/source/ProgressiveTimestampsAndWatermarks.java

```diff
@@ -273,6 +322,16 @@ void emitPeriodicWatermark() {
             }
             watermarkMultiplexer.onPeriodicEmit();
         }
+
+        public void pauseOrResumeSplits(
+                Collection<String> splitsToPause, Collection<String> splitsToResume) {
+            for (String splitId : splitsToPause) {
+                inputActivityClocks.get(splitId).markBlocked();
+            }
+            for (String splitId : splitsToResume) {
+                inputActivityClocks.get(splitId).markUnblocked();
+            }
```

Comment:
IIUC, there's no ordering guarantee between `pauseOrResumeSplits` and `releaseOutputForSplit`.
So if `releaseOutputForSplit` comes first then this method might produce NPE.
Or am I missing something?

Reply:
AFAIU there should be such ordering guarantee. Output is released (~ `closed` ?) only if split for which it was created has finished.

Reply:
Yes, but the `WatermarkAlignmentEvent` is [sent](https://github.com/apache/flink/blob/e126f382a1f52e228a8a361b9a828a92269b6fa3/flink-runtime/src/main/java/org/apache/flink/runtime/source/coordinator/SourceCoordinator.java#L214) from `SourceCoordinator` (via RPC), so I _**guess**_  these paths aren't synchronized. I might be wrong though because I'm not familiar with this part of code.

Reply:
But that RPC contains only `long maxWatermark` value. So yes, that value might be outdated due to race conditions. But splits are purely handled in the source operator's thread, which should take care of internal consistency of it's fields. 

In other words, when RPC with `WatermarkAlignmentEvent` is received, it will be processed in the source operator's thread. `SourceOperator` will then iterate over it's currently active (not closed/released) splits in `checkSplitWatermarkAlignment()` method, and call `pauseOrResumeSplits()` appropriately. On the other hand, outputs/splits are released also from `SourceOperator`'s thread, during `org.apache.flink.streaming.api.operators.SourceOperator#emitNext`. So there should be no opportunity for a race condition between those two.

Reply:
>  remove the split from inputActivityClocks(source coordinator is not aware of this)

It doesn't matter. `SourceOperator` is responsible for keeping consistent state. When it comes to watermark alignment I think `SourceCoordinator` doesn't even know what split is, so it doesn't matter that some split has finished or not.

Having said that, I am now a bit worried that nothing is ever removed from `org.apache.flink.streaming.api.operators.SourceOperator#splitCurrentWatermarks` 🤔

Reply:
> Having said that, I am now a bit worried that nothing is ever removed from org.apache.flink.streaming.api.operators.SourceOperator#splitCurrentWatermarks 🤔

I have fixed that in my last commit: https://github.com/apache/flink/pull/25167/commits/d06a0c875fe621a83e462b15a6552b027d50d1ca

Reply:
Thanks, this change should solve the problem.